test/integration/health_check_integration_test.cc

```diff
@@ -253,6 +253,72 @@ TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointTimeoutHttp) {
   EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
 }
 
+// Tests that health checking gracefully handles a NO_ERROR GOAWAY from the upstream.
+TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointGoAway) {
+  initialize();
+
+  // GOAWAY doesn't exist in HTTP1.
+  if (upstream_protocol_ == FakeHttpConnection::Type::HTTP1) {
+    return;
+  }
+
+  const uint32_t cluster_idx = 0;
+  initHttpHealthCheck(cluster_idx);
+
+  // Send a GOAWAY with NO_ERROR and then a 200. The health checker should allow the request
+  // to finish despite the GOAWAY.
+  clusters_[cluster_idx].host_fake_connection_->raiseGoAway();
+  clusters_[cluster_idx].host_stream_->encodeHeaders(
+      Http::TestResponseHeaderMapImpl{{":status", "200"}}, true);
+
+  test_server_->waitForCounterGe("cluster.cluster_1.health_check.success", 1);
+  EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.success")->value());
+  EXPECT_EQ(0, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
+}
+
+// Tests that health checking properly handles a GOAWAY with an error, followed by a reset.
+TEST_P(HttpHealthCheckIntegrationTest, SingleEndpointGoAwayErrorAndReset) {
+  initialize();
+
+  // GOAWAY doesn't exist in HTTP1.
+  if (upstream_protocol_ == FakeHttpConnection::Type::HTTP1) {
+    return;
+  }
+
+  const uint32_t cluster_idx = 0;
+  initHttpHealthCheck(cluster_idx);
+
+  // Send a GOAWAY with an error. The health checker should treat this as an
+  // error and cancel the request.
+  clusters_[cluster_idx].host_fake_connection_->raiseProtocolError();
+  ASSERT_TRUE(clusters_[cluster_idx].host_fake_connection_->waitForDisconnect(
+      std::chrono::milliseconds(1000)));
```

Comment:
@mattklein123 I added this after your last round of comments, and confirmed this fails if I comment out `raiseProtocolError()`.

Reply:
I would use the default timeout, 15s or whatever.